File: scripts/validate_raw_docs.py
#!/usr/bin/env python3
"""
validate_raw_docs.py — Cơ chế tự động phát hiện file sai domain trong 00_raw_docs/.

Quét tất cả file .md trong thư mục 00_raw_docs/ và các subdir, kiểm tra:
1. File trong subdir có frontmatter domain: khớp với tên thư mục cha không.
2. File ở root (trừ RULE.md) có được khai báo trong legacy exceptions không.
3. Thư mục con có RULE.md riêng không.

Usage:
    python3 scripts/validate_raw_docs.py           # Check + báo cáo
    python3 scripts/validate_raw_docs.py --fix     # Tự động sửa (nếu an toàn)
"""

import os
import sys
import re
import yaml
from pathlib import Path

VAULT_ROOT = Path(__file__).resolve().parent.parent
RAW_DOCS_DIR = VAULT_ROOT / "vault" / "00_raw_docs"

# Legacy exceptions: các file ở root 00_raw_docs/ được phép tồn tại
LEGACY_FILES = {"RULE.md"}

# Mapping domain slug → tên thư mục con
def domain_from_dirname(dirname: str) -> str:
    """Chuyển tên thư mục thành domain slug canonical."""
    return dirname.replace("_", "-")


def extract_frontmatter(filepath: Path) -> tuple[dict | None, list[str]]:
    """Đọc YAML frontmatter từ file .md. Trả về (data, lines)."""
    try:
        content = filepath.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as e:
        return None, [f"LỖI: Không thể đọc file: {e}"]

    errors = []

    # Kiểm tra có frontmatter không
    if not content.startswith("---"):
        errors.append(f"CẢNH BÁO: File '{filepath.name}' không có YAML frontmatter.")
        return None, errors

    parts = content.split("---")
    if len(parts) < 3:
        errors.append(f"CẢNH BÁO: File '{filepath.name}' có frontmatter không đúng định dạng.")
        return None, errors

    yaml_block = parts[1]
    try:
        data = yaml.safe_load(yaml_block)
    except yaml.YAMLError as e:
        errors.append(f"LỖI: File '{filepath.name}' có YAML lỗi: {e}")
        return None, errors

    if not isinstance(data, dict):
        errors.append(f"CẢNH BÁO: File '{filepath.name}' có frontmatter rỗng hoặc không phải dict.")
        return None, errors

    return data, errors


def validate_subdir(subdir: Path, dirname: str, fix: bool = False) -> list[str]:
    """Kiểm tra tất cả file .md trong một subdir."""
    domain_slug = domain_from_dirname(dirname)
    reports = []
    ruletxt_path = subdir / "RULE.md"

    # Kiểm tra RULE.md
    if not ruletxt_path.exists():
        reports.append(f"❌ '{dirname}/' thiếu RULE.md riêng!")

    for fpath in sorted(subdir.iterdir()):
        if fpath.name == "RULE.md":
            continue
        if not fpath.suffix == ".md":
            continue

        data, errs = extract_frontmatter(fpath)
        reports.extend(errs)

        # Kiểm tra field 'domain'
        if data and "domain" not in data:
            reports.append(f"⚠️ '{dirname}/{fpath.name}' thiếu trường 'domain: ' trong frontmatter.")
        elif data and data.get("domain") != domain_slug:
            old_domain = data.get("domain", "(không có)")
            reports.append(
                f"❌ '{dirname}/{fpath.name}' có domain='{old_domain}' "
                f"— phải là '{domain_slug}'."
            )

        # Kiểm tra field 'status'
        if data and "status" not in data:
            reports.append(f"⚠️ '{dirname}/{fpath.name}' thiếu trường 'status: ' trong frontmatter.")

    return reports


def validate_root(fix: bool = False) -> list[str]:
    """Kiểm tra các file ở root 00_raw_docs/."""
    reports = []

    for fpath in sorted(RAW_DOCS_DIR.iterdir()):
        if not fpath.is_file() or not fpath.suffix == ".md":
            continue
        if fpath.name in LEGACY_FILES:
            continue

        # File legacy (sutta-mn-*, lecture-*, v.v.) — chỉ cảnh báo nhẹ
        reports.append(
            f"ℹ️  '{fpath.name}' là file legacy ở root 00_raw_docs/. "
            "Domain mới PHẢI đặt trong subdir."
        )

    return reports


def main():
    fix = "--fix" in sys.argv

    print("=" * 60)
    print("🔍 00_raw_docs/ Validation Report")
    print("=" * 60)
    print()

    all_reports = []
    all_errors = []
    all_warnings = []

    # Kiểm tra root
    root_reports = validate_root(fix=fix)
    all_reports.extend(root_reports)

    # Kiểm tra từng subdir
    for entry in sorted(RAW_DOCS_DIR.iterdir()):
        if not entry.is_dir():
            continue
        dirname = entry.name
        reports = validate_subdir(entry, dirname, fix=fix)
        all_reports.extend(reports)

    # Phân loại báo cáo
    for r in all_reports:
        if r.startswith("❌"):
            all_errors.append(r)
        elif r.startswith("⚠️") or r.startswith("ℹ️"):
            all_warnings.append(r)

    # In kết quả
    if all_errors:
        print("## ❌ Lỗi cần sửa")
        for e in all_errors:
            print(f"  {e}")
        print()

    if all_warnings:
        print("## ⚠️ Cảnh báo")
        for w in all_warnings:
            print(f"  {w}")
        print()

    if not all_errors and not all_warnings:
        print("✅ Không phát hiện vấn đề gì.")
    else:
        print(f"📊 Tổng kết: {len(all_errors)} lỗi, {len(all_warnings)} cảnh báo.")

    return 0 if not all_errors else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/validate_raw_docs.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import matter from 'gray-matter';

const VAULT_ROOT = path.resolve(import.meta.dirname, '..');
const RAW_DOCS_DIR = path.join(VAULT_ROOT, 'vault', '00_raw_docs');
const LEGACY_FILES = new Set(['RULE.md']);

function domainFromDirname(dirname: string): string {
  return dirname.replace(/_/g, '-');
}

function validateSubdir(subdirPath: string, dirname: string): string[] {
  const domainSlug = domainFromDirname(dirname);
  const reports: string[] = [];
  const rulePath = path.join(subdirPath, 'RULE.md');

  if (!fs.existsSync(rulePath)) {
    reports.push(`❌ '${dirname}/' thiếu RULE.md riêng!`);
  }

  const files = fs.readdirSync(subdirPath).sort();
  for (const file of files) {
    if (file === 'RULE.md' || !file.endsWith('.md')) continue;

    const filePath = path.join(subdirPath, file);
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const parsed = matter(content);
      const data = parsed.data || {};

      if (!data.domain) {
        reports.push(`⚠️ '${dirname}/${file}' thiếu trường 'domain: ' trong frontmatter.`);
      } else if (data.domain !== domainSlug) {
        reports.push(`❌ '${dirname}/${file}' có domain='${data.domain}' — phải là '${domainSlug}'.`);
      }

      if (!data.status) {
        reports.push(`⚠️ '${dirname}/${file}' thiếu trường 'status: ' trong frontmatter.`);
      }
    } catch (e: any) {
      reports.push(`❌ Lỗi đọc file '${dirname}/${file}': ${e.message}`);
    }
  }

  return reports;
}

function validateRoot(): string[] {
  const reports: string[] = [];
  if (!fs.existsSync(RAW_DOCS_DIR)) return reports;

  const entries = fs.readdirSync(RAW_DOCS_DIR).sort();
  for (const entry of entries) {
    const fullPath = path.join(RAW_DOCS_DIR, entry);
    if (fs.statSync(fullPath).isFile() && entry.endsWith('.md')) {
      if (LEGACY_FILES.has(entry)) continue;
      reports.push(`ℹ️  '${entry}' là file legacy ở root 00_raw_docs/. Domain mới PHẢI đặt trong subdir.`);
    }
  }
  return reports;
}

function main() {
  console.log('='.repeat(60));
  console.log('🔍 00_raw_docs/ Validation Report (TypeScript)');
  console.log('='.repeat(60));
  console.log();

  const allReports: string[] = [];
  allReports.push(...validateRoot());

  if (fs.existsSync(RAW_DOCS_DIR)) {
    const entries = fs.readdirSync(RAW_DOCS_DIR).sort();
    for (const entry of entries) {
      const fullPath = path.join(RAW_DOCS_DIR, entry);
      if (fs.statSync(fullPath).isDirectory()) {
        allReports.push(...validateSubdir(fullPath, entry));
      }
    }
  }

  const errors = allReports.filter(r => r.startsWith('❌'));
  const warnings = allReports.filter(r => r.startsWith('⚠️') || r.startsWith('ℹ️'));

  if (errors.length > 0) {
    console.log('## ❌ Lỗi cần sửa');
    errors.forEach(e => console.log(`  ${e}`));
    console.log();
  }

  if (warnings.length > 0) {
    console.log('## ⚠️ Cảnh báo');
    warnings.forEach(w => console.log(`  ${w}`));
    console.log();
  }

  if (errors.length === 0 && warnings.length === 0) {
    console.log('✅ Không phát hiện vấn đề gì.');
  } else {
    console.log(`📊 Tổng kết: ${errors.length} lỗi, ${warnings.length} cảnh báo.`);
  }

  if (errors.length > 0) {
    process.exit(1);
  }
}

main();
